### PR TITLE
[onert] Add unit tests for training mean op

### PR DESCRIPTION
**Pull request by ragmani**
This commit adds some unit tests for mean op to validate training.

ONE-DCO-1.0-Signed-off-by: ragmani <ragmani0216@gmail.com>

### CODE REVIEW

File: tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Mean.test.cc [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    float learning_rate = 0.01f;
+    int32_t batch_size = 1;
+    cgen.addTrainInfo({circle::Optimizer::Optimizer_SGD, learning_rate,
+                       circle::LossFn::LossFn_MEAN_SQUARED_ERROR,
+                       circle::LossReductionType::LossReductionType_SumOverBatchSize, batch_size});
+
+    _context = std::make_unique<GenModelTrainContext>(cgen.finish());
+    _context->addTrainCase(uniformTCD<float>({{{1, 3}}, {{2, 1}}}, // inputs
+                                             {{{5}}, {{3}}},       // expected
+                                             {13.3691f}            // loss
```
Comment (ragmani) [inline tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Mean.test.cc:48]: The result of executing the same model with the same data on tensorflow:
```
Epoch 1/5
2/2 [==============================] - 0s 4ms/step - loss: 16.7778 - categorical_accuracy: 1.0000
Epoch 2/5
2/2 [==============================] - 0s 1ms/step - loss: 15.5544 - categorical_accuracy: 1.0000
Epoch 3/5
2/2 [==============================] - 0s 1ms/step - loss: 14.4203 - categorical_accuracy: 1.0000
Epoch 4/5
2/2 [==============================] - 0s 2ms/step - loss: 13.3691 - categorical_accuracy: 1.0000
Epoch 5/5
2/2 [==============================] - 0s 1ms/step - loss: 12.3947 - categorical_accuracy: 1.0000
```

File: tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Mean.test.cc [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+                                             ));
+
+    _context->setBackends({"train"});
+    _context->setEpoch(4);
```
Comment (zetwhite) [inline tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Mean.test.cc:52]: "setting epoch 4" is intentional? 
I first thought it was mistyping, Because the log of TensorFlow (https://github.com/Samsung/ONE/pull/13189/files#r1639240653)  runs with `epoch=5`.

File: tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Mean.test.cc [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+    SUCCEED();
+  }
+}
+
+TEST_F(GenModelTrain, neg_NonTrainableOps_Mean_InvalidShape)
+{
+  CirclePlusGen cgen;
+  int input = cgen.addTensor({{1, 8}, circle::TensorType::TensorType_FLOAT32});
+  int axis = cgen.addTensor({{2}, circle::TensorType::TensorType_INT32});
```
Comment (zetwhite) [inline tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Mean.test.cc:62]: (suggestion) 

```suggestion
  // Invalid axis: axis should be smaller than the rank of the input
  int axis = cgen.addTensor({{2}, circle::TensorType::TensorType_INT32});
```

File: tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Mean.test.cc (hunk 2 above)
Comment (ragmani) [reply tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Mean.test.cc:52]: I didn't have a special intention. It means the last loss value is the result after execution for the number of 4(epochs). We can change it later whenever.

File: tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Mean.test.cc [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    SUCCEED();
+  }
+}
+
+TEST_F(GenModelTrain, neg_NonTrainableOps_Mean_InvalidShape)
+{
+  CirclePlusGen cgen;
+  int input = cgen.addTensor({{1, 8}, circle::TensorType::TensorType_FLOAT32});
+  // Invalid axis: axis should be smaller than the rank of the input
+  int axis = cgen.addTensor({{2}, circle::TensorType::TensorType_INT32});
```
Comment (zetwhite) [inline tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Mean.test.cc:63]:
```suggestion
  int axis = cgen.addTensor({{2}, circle::TensorType::TensorType_INT32});
```
@ragmani Sorry for confusion. 
I misunderstood the code. `{2}` is not the value of the axis, it is the size(rank) of the axis.
So this comment I left before looks wrong.

Comment (zetwhite) [reply tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Mean.test.cc:63]: By the way, I was trying to add a simple comment about why this test case is expected to fail. 
I thought that a simple comment would help to quickly catch the code.

File: tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Mean.test.cc (hunk 2 above)
Comment (ragmani) [reply tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Mean.test.cc:52]: I change the comment to "// Invalid shape: input and output's ranks should be equal if keep_dims is true"

File: tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Mean.test.cc (hunk 4 above)
Comment (ragmani) [reply tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Mean.test.cc]: I change the comment to "// Invalid shape: input and output's ranks should be equal if keep_dims is true"

File: tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Mean.test.cc [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+  _context = std::make_unique<GenModelTrainContext>(cgen.finish());
+  _context->setBackends({"train"});
+  _context->expectFailCompile();
+}
+
+TEST_F(GenModelTrain, neg_NonTrainableOps_Mean_InvalidType)
+{
+  CirclePlusGen cgen;
+  int input = cgen.addTensor({{1, 8}, circle::TensorType::TensorType_INT32});
```
Comment (zetwhite) [inline tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Mean.test.cc:83]:
```suggestion
  // Invalid  type: input tensor type should be FLOAT32 
  int input = cgen.addTensor({{1, 8}, circle::TensorType::TensorType_INT32});
```

I referred here : https://github.com/Samsung/ONE/blob/be96c99f1a1e2a301aa3a6ddd3b9601ce35a6aa7/runtime/onert/backend/cpu/ops/MeanLayer.cc#L64-L73

File: tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Mean.test.cc (hunk 4 above)
Comment (zetwhite) [reply tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Mean.test.cc]: Thank you :smile:

**Review by zetwhite — DISMISSED** [review]
LGTM

File: tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Mean.test.cc (hunk 5 above)
Comment (ragmani) [reply tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Mean.test.cc:83]: I think it may be redundant, but it's not bad with the comment.

**Review by zetwhite — DISMISSED** [review]

**Review by Aeren1564 — DISMISSED** [review]
LGTM!

File: tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Mean.test.cc [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    cgen.addOperatorFullyConnected({{input, weight, bias}, {fc_output}});
+    bool keep_dims = true;
+    cgen.addOperatorMean({{fc_output, axis}, {output}}, keep_dims);
+    cgen.setInputsAndOutputs({input}, {output});
+
+    float learning_rate = 0.01f;
+    int32_t batch_size = 1;
+    cgen.addTrainInfo({circle::Optimizer::Optimizer_SGD, learning_rate,
+                       circle::LossFn::LossFn_MEAN_SQUARED_ERROR,
+                       circle::LossReductionType::LossReductionType_SumOverBatchSize, batch_size});
```
Comment (ragmani) [inline tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Mean.test.cc:43]:
```suggestion
                       circle::LossReductionType::LossReductionType_SumOverBatchSize, batch_size});
    cgen.markAllOpsAsTrainable();
```

File: tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Mean.test.cc [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+  int out = cgen.addTensor({{1, 2, 3}, circle::TensorType::TensorType_FLOAT32});
+  bool keep_dims = true;
+  cgen.addOperatorMean({{input, axis}, {out}, keep_dims}, circle::ActivationFunctionType_NONE);
+  cgen.setInputsAndOutputs({input, axis}, {out});
+
+  float learning_rate = 0.01f;
+  int32_t batch_size = 1;
+  cgen.addTrainInfo({circle::Optimizer::Optimizer_SGD, learning_rate,
+                     circle::LossFn::LossFn_MEAN_SQUARED_ERROR,
+                     circle::LossReductionType::LossReductionType_SumOverBatchSize, batch_size});
```
Comment (ragmani) [inline tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Mean.test.cc:73]:
```suggestion
                     circle::LossReductionType::LossReductionType_SumOverBatchSize, batch_size});
  cgen.markAllOpsAsTrainable();
```

File: tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Mean.test.cc [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+  int out = cgen.addTensor({{1, 1}, circle::TensorType::TensorType_FLOAT32});
+  bool keep_dims = true;
+  cgen.addOperatorMean({{input, axis}, {out}, keep_dims}, circle::ActivationFunctionType_NONE);
+  cgen.setInputsAndOutputs({input, axis}, {out});
+
+  float learning_rate = 0.01f;
+  int32_t batch_size = 1;
+  cgen.addTrainInfo({circle::Optimizer::Optimizer_SGD, learning_rate,
+                     circle::LossFn::LossFn_MEAN_SQUARED_ERROR,
+                     circle::LossReductionType::LossReductionType_SumOverBatchSize, batch_size});
```
Comment (ragmani) [inline tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Mean.test.cc:95]:
```suggestion
                     circle::LossReductionType::LossReductionType_SumOverBatchSize, batch_size});
  cgen.markAllOpsAsTrainable();
```

**Comment by ragmani** [comment]
added `cgen.markAllOpsAsTrainable();`

**Review by jyoungyun — APPROVED** [review]
LGTM
